hudi-client/hudi-spark-client/src/main/java/org/apache/hudi/table/action/commit/UpsertPartitioner.java

```diff
@@ -185,21 +185,26 @@ private void assignInserts(WorkloadProfile profile, HoodieEngineContext context)
 
         // first try packing this into one of the smallFiles
         for (SmallFile smallFile : smallFiles) {
-          long recordsToAppend = Math.min((config.getParquetMaxFileSize() - smallFile.sizeBytes) / averageRecordSize,
-              totalUnassignedInserts);
-          if (recordsToAppend > 0 && totalUnassignedInserts > 0) {
-            // create a new bucket or re-use an existing bucket
-            int bucket;
-            if (updateLocationToBucket.containsKey(smallFile.location.getFileId())) {
-              bucket = updateLocationToBucket.get(smallFile.location.getFileId());
-              LOG.info("Assigning " + recordsToAppend + " inserts to existing update bucket " + bucket);
-            } else {
-              bucket = addUpdateBucket(partitionPath, smallFile.location.getFileId());
-              LOG.info("Assigning " + recordsToAppend + " inserts to new update bucket " + bucket);
+          if (totalUnassignedInserts > 0) {
+            long recordsToAppend = Math.min((config.getParquetMaxFileSize() - smallFile.sizeBytes) / averageRecordSize,
+                totalUnassignedInserts);
+            if (recordsToAppend > 0) {
+              // create a new bucket or re-use an existing bucket
+              int bucket;
+              if (updateLocationToBucket.containsKey(smallFile.location.getFileId())) {
+                bucket = updateLocationToBucket.get(smallFile.location.getFileId());
+                LOG.info("Assigning " + recordsToAppend + " inserts to existing update bucket " + bucket);
+              } else {
+                bucket = addUpdateBucket(partitionPath, smallFile.location.getFileId());
+                LOG.info("Assigning " + recordsToAppend + " inserts to new update bucket " + bucket);
+              }
+              bucketNumbers.add(bucket);
+              recordsPerBucket.add(recordsToAppend);
+              totalUnassignedInserts -= recordsToAppend;
             }
-            bucketNumbers.add(bucket);
-            recordsPerBucket.add(recordsToAppend);
-            totalUnassignedInserts -= recordsToAppend;
+          } else {
```

Comment:
if the main thing you want to fix is to break out early. Can we do it after L203, with a if check, without nesting this 1 level under again in if block?

```Java 
totalUnassignedInserts -= recordsToAppend;
if (totalUnassignedInserts <= 0) {
  break;
}
```

Reply:
> if the main thing you want to fix is to break out early. Can we do it after L203, with a if check, without nesting this 1 level under again in if block?
> 
> ```java
> totalUnassignedInserts -= recordsToAppend;
> if (totalUnassignedInserts <= 0) {
>   break;
> }
> ```

Yes, I want to break it early. done